server/persistence-plugin-mode/persistence-plugin-mode.md

```diff
@@ -0,0 +1,89 @@
+# Proposal: Persistence Plugin Model
+
+Author: Wenquan Xing
+
+Last updated: Oct 2020
+
+Discussion at TODO
+
+## Abstract
+
+This document describe the existing & target solution towards supporting multiple storage backends for Temporal server.
+
+## Background
+
+Existing implementation of Temporal server already provide decent abstraction between business logic vs persistence logic. In fact, lots of work have already be done making the SQL persistence layer pluggable, i.e. [link](https://github.com/temporalio/temporal/tree/master/common/persistence/sql/sqlplugin).
+
+Although huge efforts have been dedicated to making the persistence layer standalone and extensible, there are still areas for improvement, e.g. better test coverage with reduced complexity & pluggable NoSQL persistence layer.
+
+## Proposal
+
+### Desired state
+
+Most of the existing high level code / logic structure will be maintained, to limit the potential blast radius as well as for project management.
+
+All plugin model changes will be hidden behind the persistence interface [link](https://github.com/temporalio/temporal/blob/master/common/persistence/persistenceInterface.go).
+
+Plugins will be categorized into 2 buckets: NoSQL and SQL. Please see the component diagram and definition for details:
+
+![image alt text](layout.png)
+
+Definition:
+
+* Business logic
+* Persistence interface: persistence APIs for storage of workflow history events; mutable state changes; async tasks.
+* Persistence test: overall test logic against generic persistence layer.
+* SQL persistence: generic SQL style persistence
+  * SQL persistence test: SQL specific tests.
+  * SQL persistence common logic: common logic which operates on top of SQL persistence drivers.
+  * SQL persistence driver: MySQL; Postgres.
```

Comment:
It is not Postgres. Can you please consistenly use the correct name `PostgreSQL`?